src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,1763 @@
+                let is_lm = tag.eq_ignore_ascii_case(LAST_MODIFIED);
+                if is_etag || is_lm {
+                    // Append a new line to each so we can easily prepend to the index file.
+                    let mut s = String::with_capacity(value.len() + 1);
+                    s.push_str(value);
+                    s.push('\n');
+                    if is_etag {
+                        etag = Some(s);
+                    } else if is_lm {
+                        last_modified = Some(s);
+                    }
+                }
+            }
+
+            true
+        })?;
+
+        transfer
+            .perform()
+            .chain_err(|| format!("failed to fetch index file `{}`", path.display()))?;
+        drop(transfer);
+
+        // Avoid the same conditional headers being sent in future re-uses of the `Easy` client.
+        let mut list = List::new();
+        list.append("If-Modified-Since:")?;
+        list.append("If-None-Match:")?;
+        handle.http_headers(list)?;
+        let response_code = handle.response_code()?;
+        drop(handle);
+
+        debug!("index file downloaded with status code {}", response_code,);
+
+        // Make sure we don't double-check the file again if it's loaded again.
+        assert!(
+            self.fresh.insert(path.to_path_buf()),
+            "downloaded the index file `{}` twice",
+            path.display(),
+        );
+
+        match response_code {
+            200 => {}
+            304 => {
+                // Not Modified response.
+                let (_, _, bytes) =
+                    was.expect("conditional request response implies we have local index file");
+                return data(bytes);
+            }
+            404 | 410 | 451 => {
+                // The crate was deleted from the registry.
+                if was.is_some() {
+                    // Make sure we delete the local index file.
+                    debug!("crate {} was deleted from the registry", path.display());
+                    paths::remove_file(&pkg)?;
+                }
+                anyhow::bail!("crate has been deleted from the registry");
+            }
+            code => {
+                anyhow::bail!("server returned unexpected HTTP status code {}", code);
+            }
+        }
+
+        paths::create_dir_all(pkg.parent().expect("pkg is a file"))?;
+        let mut file = paths::create(&root.join(path))?;
+        file.write_all(etag.as_deref().unwrap_or("\n").as_bytes())?;
+        file.write_all(last_modified.as_deref().unwrap_or("\n").as_bytes())?;
+        file.write_all(&contents)?;
+        file.flush()?;
+        data(&contents)
+    }
+
+    fn config(&mut self) -> CargoResult<Option<RegistryConfig>> {
+        debug!("loading config");
+        self.prepare()?;
+        self.config.assert_package_cache_locked(&self.index_path);
+        let mut config = None;
+        self.load(Path::new(""), Path::new("config.json"), &mut |json| {
```

Comment:
Finally figured out <https://github.com/rust-lang/cargo/pull/8890#issuecomment-737481586>, this is reading and writing `config.json` in the current working directory, so if you have multiple `HttpRegistry` they will overwrite each others cached configs.

Reply:
Ah, good catch! That's what I get for not leaving a `TODO:` for myself during prototyping. Should be fixed now! Thanks for digging!